src/compiler/transformers/jsx.ts

```diff
@@ -24,7 +26,458 @@ namespace ts {
+            const text = getTextOfNode(node, /*includeTrivia*/ true);
+            let result: string = undefined;
+            let firstNonWhitespace = 0;
+            let lastNonWhitespace = -1;
+
+            // JSX trims whitespace at the end and beginning of lines, except that the
+            // start/end of a tag is considered a start/end of a line only if that line is
+            // on the same line as the closing tag. See examples in tests/cases/conformance/jsx/tsxReactEmitWhitespace.tsx
+            for (let i = 0; i < text.length; i++) {
+                const c = text.charCodeAt(i);
+                if (isLineBreak(c)) {
+                    if (firstNonWhitespace !== -1 && (lastNonWhitespace - firstNonWhitespace + 1 > 0)) {
+                        const part = text.substr(firstNonWhitespace, lastNonWhitespace - firstNonWhitespace + 1);
+                        result = (result ? result + "\" + ' ' + \"" : "") + part;
+                    }
+                    firstNonWhitespace = -1;
+                }
+                else if (!isWhiteSpace(c)) {
+                    lastNonWhitespace = i;
+                    if (firstNonWhitespace === -1) {
+                        firstNonWhitespace = i;
+                    }
+                }
+            }
+
+            if (firstNonWhitespace !== -1) {
+                const part = text.substr(firstNonWhitespace);
+                result = (result ? result + "\" + ' ' + \"" : "") + part;
+            }
+
+            if (result) {
+                // Replace entities like &nbsp;
+                result = result.replace(/&(\w+);/g, function(s: any, m: string) {
+                    if (entities[m] !== undefined) {
+                        return String.fromCharCode(entities[m]);
+                    }
+                    else {
+                        return s;
+                    }
+                });
+            }
+
+            return result;
+        }
+
+        function getTagName(node: JsxElement | JsxOpeningLikeElement): Expression {
+            if (node.kind === SyntaxKind.JsxElement) {
+                return getTagName((<JsxElement>node).openingElement);
+            }
+            else {
+                const name = (<JsxOpeningLikeElement>node).tagName;
+                if (isIdentifier(name) && isIntrinsicJsxName(name.text)) {
+                    return createLiteral(name.text);
+                }
+                else {
+                    return createExpressionFromEntityName(name);
+                }
+            }
         }
+
+        /**
+         * Emit an attribute name, which is quoted if it needs to be quoted. Because
+         * these emit into an object literal property name, we don't need to be worried
+         * about keywords, just non-identifier characters
+         */
+        function getAttributeName(node: JsxAttribute): StringLiteral | Identifier {
+            const name = node.name;
+            if (/[A-Za-z_]+[\w*]/.test(name.text)) {

```

Comment:
1. why is `\w*` in brackets? Right now it matches a single word-character or a star, not 0-or-more word-characters.
2. this matches text anywhere in the string, so it looks like an identifier in the middle of other stuff will match. For example `"hi there this is dash-separated"` Is this why it's quoted? It seems like we'll end up quoting way too many things.